### PR TITLE
HDDS-5909. EC: Onboard EC into upgrade framework

### PR DESCRIPTION
**Pull request by fapifta**
## What changes were proposed in this pull request?

The change introduces the OM and HDDS LayoutFeature for EC. In order for EC to work, there are no upgrade actions necessary so none are defined, EC does work based on the normal facilities, and solves the erasure coded storage of data solely on the client side. The server side just gets knowledge about the fact that data inside an object is erasure coded with the ReplicationConfig.

The change also introduces a new ClientVersion to mark that the client side supports erasure coded storage, also fixes a few APIDoc typos where ClientVersion remained ClientVersions during the rename of the class.

The change also adds request and response validation to ensure the following backward compatibility and upgrade related questions:
- During an upgrade until the cluster is finalized, there aren't any way to create a key/bucket/container/pipeline with ECReplicationConfig.
- After an upgrade a client is not served with any type of object that has an ECReplicationconfig that the old client would not understand.
The change affects SCM and OM, so that SCM Admin and SCM Container Location requests are also disabled alongside with any requests that the OM ClientProtocol provides.

## What is the link to the Apache JIRA

https://issues.apache.org/jira/browse/HDDS-5909

## How was this patch tested?

As the logic added is so simple, I did not added new tests for the request/response validation so far, current tests should work without any modification, while in order to add compatibility tests for the client and server side I believe we need a bigger effort to extend what we already have for compatibility alongside with the development related to HDDS-6390

### CODE REVIEW

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/protocolPB/OzoneManagerRequestHandler.java
Comment (errose28) [inline hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/protocolPB/OzoneManagerRequestHandler.java:730]: nit. Missing space in concatenation.

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/protocolPB/OzoneManagerRequestHandler.java
Comment (errose28) [inline hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/protocolPB/OzoneManagerRequestHandler.java:769]: nit. Missing space in concatenation.

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/protocolPB/OzoneManagerRequestHandler.java
Comment (errose28) [inline hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/protocolPB/OzoneManagerRequestHandler.java:391]: This will block the request for all old clients, but a few versions in the future, clients can be old but still have EC support.

I think we could get the client version to this method by having `RequestValidations#validateResponse` (and `RequestValidations#validateRequest` as well) take the context as a parameter which they can forward to the annotated methods, instead of all the annotated methods receiving the same pre-constructed context. `OzoneManagerProtocolServerSideTranslatorPB#submitRequest` should have enough information to construct this context when invoking these methods.

Let me know what you think of this approach, or if you have a better idea.

**Review by errose28 — COMMENTED** [review]
Thanks for working on this @fapifta. Some comments and questions in line.

I am not very familiar with the EC implementation (yet) so someone else might be better to double check that the requests blocked off here are the correct ones. Also, do we need to block `AllocateScmBlock` request in `ScmBlockLocationProtocolServerSideTranslatorPB`?

Some minimal testing would be nice as well. One simple option is to add an EC key create to the pre-finalize and finalize stages of the upgrade/downgrade acceptance tests, and check that it fails in pre-finalize and passes in finalize. The xcompat acceptance tests should allow similar functionality to test with old clients. It is also easy, although slow, to create a pre-finalized mini ozone cluster. If the tests there are not likely to break they could be annotated with the new `@Slow` annotation to not run them in the main CI.

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/protocolPB/OzoneManagerRequestHandler.java
Comment (kerneltime) [inline hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/protocolPB/OzoneManagerRequestHandler.java:446]: Same as above, the client's actual version matters as much as the relative version.

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/protocolPB/OzoneManagerRequestHandler.java
Comment (kerneltime) [inline hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/protocolPB/OzoneManagerRequestHandler.java:479]: Why disallow the list bucket command? Ozone CLI can be an older version and it should be able to list all the buckets.

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/protocolPB/OzoneManagerRequestHandler.java
Comment (errose28) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/protocolPB/OzoneManagerRequestHandler.java:395]: Right, this looks like a new optional field added to the BucketInfo proto. If the old client gets it I would think protobuf would ignore the field and not throw an error.

File: hadoop-hdds/common/src/main/java/org/apache/hadoop/ozone/ClientVersion.java
Comment (fapifta) [reply hadoop-hdds/common/src/main/java/org/apache/hadoop/ozone/ClientVersion.java]: done

File: hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/protocol/ScmBlockLocationProtocolServerSideTranslatorPB.java
Comment (fapifta) [reply hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/protocol/ScmBlockLocationProtocolServerSideTranslatorPB.java]: done, thank you for noting it, it is way much better.

File: hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/protocol/ScmBlockLocationProtocolServerSideTranslatorPB.java
Comment (fapifta) [reply hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/protocol/ScmBlockLocationProtocolServerSideTranslatorPB.java:138]: yes, absolutely. That is why I chose this implementation, as it seemed to make refactoring to that framework easier, once it is ready.

File: hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/protocol/StorageContainerLocationProtocolServerSideTranslatorPB.java
Comment (fapifta) [reply hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/protocol/StorageContainerLocationProtocolServerSideTranslatorPB.java]: yes, you are right. done.

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/file/OMDirectoryCreateRequest.java
Comment (fapifta) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/file/OMDirectoryCreateRequest.java:428]: Actually it does not mean anything as we support only the bucket level default replication configuration.

There are two reasons why I have added it here:
1. we are on the server side, and anyone can send us a request, not just our client, so a validation on this is useful in theory... in practice it may not ever gonna happen
2. if anytime in the future we decide to support specifying EC replication on a directory level like in HDFS, then this would be already here.

If neither of these seems to be useful, I am open to just remove this check anytime.

On the other hand I changed this to isallowed and reverted the condition.

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/protocolPB/OzoneManagerRequestHandler.java
Comment (fapifta) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/protocolPB/OzoneManagerRequestHandler.java]: done, thanks!

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/protocolPB/OzoneManagerRequestHandler.java
Comment (fapifta) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/protocolPB/OzoneManagerRequestHandler.java]: oh that magical copy paste... thank you! done

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/protocolPB/OzoneManagerRequestHandler.java
Comment (fapifta) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/protocolPB/OzoneManagerRequestHandler.java:391]: According to my plans, this I would like to clear before cutting the 1.3.0 release, and I would like to see the whole framework implemented at least on OM and SCM side, but probably on the client side as well.
I noticed this discrepancy also, so I took a note for my plans that I would re-formalize and publish as a design plan in the near future.

I have two competing ideas in mind for this,
one is to add the version to the annotation to mark when a validator has to run, so then the classification of the validators can be done in a per version base as the innermost structure in the registry. (this will lengthen the chain of map of map of map of lists to map of map of map of map of lists though).
the other is to add checks into the validators based on the OMRequest->version field, which is already given to all validators

I think the second one is sufficient for quite a while, and we can introduce the first one later, the first one is something that would perform and work well on the long run.

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/protocolPB/OzoneManagerRequestHandler.java
Comment (fapifta) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/protocolPB/OzoneManagerRequestHandler.java:446]: agree, see my answer to @errose28 @kerneltime. As I think we should come up with the refactor prior to 1.3.0 release and that should contain a functionality like this for sure, this changeset is now about expediting the branch merge for the EC branch by ensuring the compatibility parts of it.

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/protocolPB/OzoneManagerRequestHandler.java
Comment (fapifta) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/protocolPB/OzoneManagerRequestHandler.java:479]: As we discussed in the JIRA and offline as well, there are two things we can do here:
- filter EC replicated stuff from the result
- fail the request

If we filter, there can be two inconsistent views of the filesystem from an old and from a new client (which might be on the same user machine in two different softwares), which can lead to confusion. So it seems to be safer to just fail the request.

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/protocolPB/OzoneManagerRequestHandler.java
Comment (fapifta) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/protocolPB/OzoneManagerRequestHandler.java:395]: It is an interesting question... I haven't checked deeply what would happen...
But... I firmly believe two things can happen:
1. proto libs initialize lazily and unless someone does not reach the ECReplicationConfig it does not try to initialize it
2. proto libs read through and parse the whole message upon arrival

In the first case, we would still need to ensure that the client will never read the value, which is tedious, and have to be checked in future commits as well up until it is released, so this would be something to check in the release in order to ensure the client does not fail. (This does not seem to be failsafe, and also it is not the easiest task either to write tests for all code branches or to go over it by hand and understand the code.)
So in the first case the client might fail if there is an ECReplicationConfig somehow in the result,
While in the second case the client will fail inevitably when there is an ECReplicaitonConfig in the result.

Based on this I think it is just safer and easier to ensure that we do not serve an ECReplicationConfig from the server side.

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/protocolPB/OzoneManagerRequestHandler.java
Comment (fapifta) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/protocolPB/OzoneManagerRequestHandler.java:523]: this check is here because of the fears in my prev answer for the disallowInfoBucketResponseWithECReplicationConfig method, and the reasoning in the answer below this one by two for disallowListBucketsResponseWithECReplicationConfig.

**Comment by fapifta** [comment]
Thank you for the reviews both of you @errose28 @kerneltime 

I tried to address most of the concerns.

I see three things we need to get to a conclusion with before we might close out this PR.
1. how exactly the proto layer fails if there is an unused ECReplicationConfig in the result
2. we either need to agree that discrepancies of the fw are solved later on during the planned refactoring and extension to SCM and possibly to the client layer as well, or we add at least the version check of the client here.
3. How we want to solve the listing related question (fail, or filter), as I used an assumption about the client side failure (I had to admit it to myself as well first :D ), this can be decided after we know the answer for 1.

Based on this I will go forward, and add some minimal tests to the xcompat tests, and check how things exactly work within that environment. Based on which we will know the answer for 1, and we can decide on 3 while we will have some tests as well (I will do my best to cover all the normal fs client side changes, but I won't bother much with the admin client side if that turns out to be too complicated and if you don't mind.

I will update here about 1. later today.

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/file/OMDirectoryCreateRequest.java
Comment (errose28) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/file/OMDirectoryCreateRequest.java:428]: I think we should leave it. The directory create is done with the same KeyArgs proto, which could have the ECReplicationConfig field or EC ReplicationType set. Even if it is ignored when processing the request and would have no effect, it's probably safer with no drawbacks to leave this validator in.

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/protocolPB/OzoneManagerRequestHandler.java
Comment (errose28) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/protocolPB/OzoneManagerRequestHandler.java:395]: My understanding is that even if the OM sets DefaultReplicationConfig the old client will still parse the proto without error and ignore the extra field. I think it would be good to confirm the behavior of an old client doing list/info on an EC bucket and it looks like that is listed as a work item [here](https://github.com/apache/ozone/pull/3262#issuecomment-1088361663), so let's wait and see what the results are.

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/protocolPB/OzoneManagerRequestHandler.java
Comment (errose28) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/protocolPB/OzoneManagerRequestHandler.java:523]: KeyArgs and BucketArgs involve different proto changes. In BucketArgs, one new field is added to the proto. In KeyArgs a new field is added, but a new value was also added to the ReplicationType enum in hdds.proto, which KeyArgs stores as the field called `type`. For this reason, I think an old client would be able to handle BucketArgs with EC info, but not KeyArgs with EC info. We should test the two cases separately to make sure.

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/protocolPB/OzoneManagerRequestHandler.java
Comment (errose28) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/protocolPB/OzoneManagerRequestHandler.java:391]: I am ok with handling this in a follow-up to unblock the EC merge. Let's file a Jira and mark it as a 1.3.0 blocker if we go that route.

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/key/OMAllocateBlockRequest.java
Comment (fapifta) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/key/OMAllocateBlockRequest.java:267]: There are multiple differences in these methods, I was also thinking about how to generify and what...
The followings are different here and there:
- the exception message
- the logic of extracting whether the actual request contains ECReplicationConfig or it does not

What is similar all the time:
- the check for whether the layout feature is allowed, we discussed about this with Ethan, this check has to go to the general framework logic later on, we decided to extract it later as part of that work.
- the resultcode used.

If you look at the differences, the majority of the code lines are still differ so even if they are added as a parameter that does not make the code much simpler for the first sight... Are you ok if we iterate on this later on after the compat framework is refactored and extended to support SCM and Client side as well?

Comment (umamaheswararao) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/key/OMAllocateBlockRequest.java:267]: I don't want to blocked on this. :-)
Let's refine later if there is a possible to simplify. I am ok with that, thanks

**Comment by fapifta** [comment]
There was a question open from before:

> how exactly the proto layer fails if there is an unused ECReplicationConfig in the result

This was answered in the docker environment, and tests are also written around this part of the code, so now we know and test the behaviour.
Based on that, bucket listing is not disabled, as there the DefaultReplicationConfig field is added as optional, so old clients does not look for it, and it may or may not present, but this is fine... With that old clients can show bucket data at listing and at info calls.
Key listing has to be disabled on the other hand, as there the ReplicationFactor field is required in old clients, but it is not sent by the new server, causing a failure if a KeyInfo arrives with ECReplicationConfig...
An option would be to use ReplicationFactor.ZERO, but that was introduced with EC, so old clients would not understand that either.

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/protocolPB/OzoneManagerRequestHandler.java
Comment (fapifta) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/protocolPB/OzoneManagerRequestHandler.java:476]: Test I believe made this clear and bucket listing and info is not disabled anymore

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/protocolPB/OzoneManagerRequestHandler.java
Comment (fapifta) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/protocolPB/OzoneManagerRequestHandler.java]: Yes you were right test showed it and now bucket info and listing is enabled

File: hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/protocol/StorageContainerLocationProtocolServerSideTranslatorPB.java
Comment (errose28) [inline hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/protocol/StorageContainerLocationProtocolServerSideTranslatorPB.java:287]: Discussed offline, but we should check if the admin CLI will do retries for this (which we don't want) because the type is ServiceException.

**Review by errose28 — APPROVED** [review]
Thanks for adding the tests @fapifta. I have some minor comments but we can address those in follow Jiras if needed. I do not think they need to block EC merge.

File: hadoop-ozone/dist/src/main/compose/xcompat/test.sh
Comment (fapifta) [reply hadoop-ozone/dist/src/main/compose/xcompat/test.sh:82]: Added HDDS-6601 to track this.

File: hadoop-ozone/dist/src/main/compose/upgrade/upgrades/non-rolling-upgrade/1.1.0-1.2.0/callback.sh
Comment (fapifta) [reply hadoop-ozone/dist/src/main/compose/upgrade/upgrades/non-rolling-upgrade/1.1.0-1.2.0/callback.sh]: yepp, you are right, this file should not have any changes, I did not notice I left this line here from the first testing iterations...

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/validation/RequestValidations.java
Comment (fapifta) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/validation/RequestValidations.java]: yepp, accidentally remained on info level, I changed this only for debugging...

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/protocolPB/OzoneManagerProtocolServerSideTranslatorPB.java
Comment (fapifta) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/protocolPB/OzoneManagerProtocolServerSideTranslatorPB.java]: Yes, remained from debug, I have removed these lines.

File: hadoop-ozone/dist/src/main/compose/upgrade/upgrades/non-rolling-upgrade/1.2.1-1.3.0/callback.sh
Comment (fapifta) [reply hadoop-ozone/dist/src/main/compose/upgrade/upgrades/non-rolling-upgrade/1.2.1-1.3.0/callback.sh:1]: Sounds good, thank you!

File: hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/protocol/StorageContainerLocationProtocolServerSideTranslatorPB.java
Comment (fapifta) [reply hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/protocol/StorageContainerLocationProtocolServerSideTranslatorPB.java]: I did a quick test today, and found that if we throw a plain ServiceException from the submitRequest method here, we run into an NP in the hadoop proto layer that the SCM admin interface uses as it seems...
This is the exception on the SCM side:
`2022-04-19 09:58:14,181 [IPC Server handler 22 on default port 9860] WARN ipc.Server: IPC Server handler 22 on default port 9860, call Call#0 Retry#0 org.apache.hadoop.hdds.scm.protocol.StorageContainerLocationProtocol.submitRequest from 172.20.0.8:42982
java.lang.NullPointerException
        at org.apache.hadoop.ipc.ProtobufRpcEngine$Server.processCall(ProtobufRpcEngine.java:477)
        at org.apache.hadoop.ipc.ProtobufRpcEngine2$Server$ProtoBufRpcInvoker.call(ProtobufRpcEngine2.java:574)
        at org.apache.hadoop.ipc.ProtobufRpcEngine2$Server$ProtoBufRpcInvoker.call(ProtobufRpcEngine2.java:552)
        at org.apache.hadoop.ipc.RPC$Server.call(RPC.java:1093)
        at org.apache.hadoop.ipc.Server$RpcCall.run(Server.java:1035)
        at org.apache.hadoop.ipc.Server$RpcCall.run(Server.java:963)
        at java.base/java.security.AccessController.doPrivileged(Native Method)
        at java.base/javax.security.auth.Subject.doAs(Subject.java:423)
        at org.apache.hadoop.security.UserGroupInformation.doAs(UserGroupInformation.java:1878)
        at org.apache.hadoop.ipc.Server$Handler.run(Server.java:2966)
`
This is because the hadoop layer in ProtobufRpcEngine$Server.processCall checks the ServiceException cause and would like to add a metric for that, but the ServiceException we throw does not have a cause...

I have modified this code to throw an SCMException with INTERNAL_ERROR as the error code, and I am adding an UNSUPPORTED_OPERATION as a resultcode here, as we do not have one.
Tested the changes manually, and now on the client side I see the INTERNAL_ERROR reported with the proper error message instead of the NPE.

**Review by errose28 — APPROVED** [review]

**Comment by umamaheswararao** [comment]
Thanks a lot @fapifta for your hard work here and thank you @errose28, @adoroszlai, @kerneltime for your reviews.

**Pull request merged**
